### PR TITLE
[python-package] add type hints on Booster.set_network()

### PR DESCRIPTION
**Pull request by jameslamb**
While working on `lightgbm.dask`, I noticed that the docstring for `machines` in `Booster.set_network()` seems to be incorrect.

It says the following:

> machines : list, set or string
    Names of machines.

However, `machines` is passed to `lightgbm.basic.c_str()`, which only accepts strings: https://github.com/microsoft/LightGBM/blob/bcf443b568052eac05cb8b11b17be9926b862f8f/python-package/lightgbm/basic.py#L2359.

If you pass a set or list, you'll get an error like.

> AttributeError: 'set' object has no attribute 'encode'

```python
import lightgbm as lgb
machines = set(["10.1.1.1:12400", "10.1.1.2:12400"])
lgb.basic.c_str(machines)
```

This PR proposes updating that docstring to make it clear that only strings are supported. It also adds type hints on the other arguments in `Booster.set_network()` and a return type hint, as part of #3756.

### CODE REVIEW

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -2336,13 +2336,13 @@ def _free_buffer(self):
...
 
-    def set_network(self, machines, local_listen_port=12400,
-                    listen_time_out=120, num_machines=1):
+    def set_network(self, machines: str, local_listen_port: int = 12400,
+                    listen_time_out: int = 120, num_machines: int = 1) -> "Booster":
         """Set the network configuration.
 
         Parameters
         ----------
-        machines : list, set or string
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:2345]: I think adding support for behavior from old docstring will be better.
https://github.com/microsoft/LightGBM/blob/bcf443b568052eac05cb8b11b17be9926b862f8f/python-package/lightgbm/basic.py#L2203-L2208

Comment (jameslamb) [reply python-package/lightgbm/basic.py:2345]: alright no problem

Comment (StrikerRUS) [reply python-package/lightgbm/basic.py]: Thanks!

**Comment by StrikerRUS** [comment]
Hmm, looks like one more random Dask error in tests.

**Comment by jameslamb** [comment]
pretty strange! here's the full logs

```text
__ test_network_params_not_required_but_respected_if_given[array-regression] ___

client = <Client: 'tcp://127.0.0.1:41279' processes=2 threads=2, memory=236.37 GB>
task = 'regression', output = 'array', listen_port = 13030

    @pytest.mark.parametrize('task', tasks)
    @pytest.mark.parametrize('output', data_output)
    def test_network_params_not_required_but_respected_if_given(client, task, output, listen_port):
        if task == 'ranking' and output == 'scipy_csr_matrix':
            pytest.skip('LGBMRanker is not currently tested on sparse matrices')
    
        if task == 'ranking':
            _, _, _, _, dX, dy, _, dg = _create_ranking_data(
                output=output,
                group=None,
                chunk_size=10,
            )
        else:
            _, _, _, dX, dy, _ = _create_data(
                objective=task,
                output=output,
                chunk_size=10,
            )
            dg = None
    
        dask_model_factory = task_to_dask_factory[task]
    
        # rebalance data to be sure that each worker has a piece of the data
        if output == 'array':
            client.rebalance()
    
        # model 1 - no network parameters given
        dask_model1 = dask_model_factory(
            n_estimators=5,
            num_leaves=5,
        )
        dask_model1.fit(dX, dy, group=dg)
        assert dask_model1.fitted_
        params = dask_model1.get_params()
        assert 'local_listen_port' not in params
        assert 'machines' not in params
    
        # model 2 - machines given
        n_workers = len(client.scheduler_info()['workers'])
        open_ports = [lgb.dask._find_random_open_port() for _ in range(n_workers)]
        dask_model2 = dask_model_factory(
            n_estimators=5,
            num_leaves=5,
            machines=",".join([
                "127.0.0.1:" + str(port)
                for port in open_ports
            ]),
        )
    
        dask_model2.fit(dX, dy, group=dg)
        assert dask_model2.fitted_
        params = dask_model2.get_params()
        assert 'local_listen_port' not in params
        assert 'machines' in params
    
        # model 3 - local_listen_port given
        # training should fail because LightGBM will try to use the same
        # port for multiple worker processes on the same machine
        dask_model3 = dask_model_factory(
            n_estimators=5,
            num_leaves=5,
            local_listen_port=listen_port
        )
        error_msg = "has multiple Dask worker processes running on it"
        with pytest.raises(lgb.basic.LightGBMError, match=error_msg):
>           dask_model3.fit(dX, dy, group=dg)
E           Failed: DID NOT RAISE <class 'lightgbm.basic.LightGBMError'>

../tests/python_package_test/test_dask.py:1086: Failed
```

That error not being raised means that the input data partitions are all on one worker.  I tried to explicitly control for that with https://github.com/microsoft/LightGBM/blob/96728a044a8cfcf6706fc265dbf9e9dfa24ccdd5/tests/python_package_test/test_dask.py#L1043-L1045 and by setting `chunk_size=10`. But I guess that is not 100% reliable?

**Review by StrikerRUS — APPROVED** [review]
LGTM, thanks!

**Comment by StrikerRUS** [comment]
@jameslamb 

> Hmm, looks like one more random Dask error in tests.

> pretty strange! 

One more time:
```
2021-03-23T11:32:00.3599665Z __ test_network_params_not_required_but_respected_if_given[array-regression] ___
2021-03-23T11:32:00.3600265Z 
2021-03-23T11:32:00.3601496Z client = <Client: 'tcp://127.0.0.1:44705' processes=2 threads=2, memory=16.70 GB>
2021-03-23T11:32:00.3602774Z task = 'regression', output = 'array', listen_port = 13030
2021-03-23T11:32:00.3603131Z 
2021-03-23T11:32:00.3603872Z     @pytest.mark.parametrize('task', tasks)
2021-03-23T11:32:00.3604785Z     @pytest.mark.parametrize('output', data_output)
2021-03-23T11:32:00.3605598Z     def test_network_params_not_required_but_respected_if_given(client, task, output, listen_port):
2021-03-23T11:32:00.3606647Z         if task == 'ranking' and output == 'scipy_csr_matrix':
2021-03-23T11:32:00.3607853Z             pytest.skip('LGBMRanker is not currently tested on sparse matrices')
2021-03-23T11:32:00.3608415Z     
2021-03-23T11:32:00.3608814Z         client.wait_for_workers(2)
2021-03-23T11:32:00.3609214Z     
2021-03-23T11:32:00.3609975Z         _, _, _, _, dX, dy, _, dg = _create_data(
2021-03-23T11:32:00.3610490Z             objective=task,
2021-03-23T11:32:00.3610940Z             output=output,
2021-03-23T11:32:00.3611381Z             chunk_size=10,
2021-03-23T11:32:00.3611829Z             group=None
2021-03-23T11:32:00.3612267Z         )
2021-03-23T11:32:00.3612672Z     
2021-03-23T11:32:00.3613153Z         dask_model_factory = task_to_dask_factory[task]
2021-03-23T11:32:00.3613612Z     
2021-03-23T11:32:00.3614090Z         # rebalance data to be sure that each worker has a piece of the data
2021-03-23T11:32:00.3615068Z         if output == 'array':
2021-03-23T11:32:00.3615633Z             client.rebalance()
2021-03-23T11:32:00.3616057Z     
2021-03-23T11:32:00.3616765Z         # model 1 - no network parameters given
2021-03-23T11:32:00.3617345Z         dask_model1 = dask_model_factory(
2021-03-23T11:32:00.3617873Z             n_estimators=5,
2021-03-23T11:32:00.3618366Z             num_leaves=5,
2021-03-23T11:32:00.3618788Z         )
2021-03-23T11:32:00.3619384Z         dask_model1.fit(dX, dy, group=dg)
2021-03-23T11:32:00.3619926Z         assert dask_model1.fitted_
2021-03-23T11:32:00.3627880Z         params = dask_model1.get_params()
2021-03-23T11:32:00.3634170Z         assert 'local_listen_port' not in params
2021-03-23T11:32:00.3635077Z         assert 'machines' not in params
2021-03-23T11:32:00.3635494Z     
2021-03-23T11:32:00.3636048Z         # model 2 - machines given
2021-03-23T11:32:00.3636712Z         n_workers = len(client.scheduler_info()['workers'])
2021-03-23T11:32:00.3637274Z         open_ports = [lgb.dask._find_random_open_port() for _ in range(n_workers)]
2021-03-23T11:32:00.3637781Z         dask_model2 = dask_model_factory(
2021-03-23T11:32:00.3638307Z             n_estimators=5,
2021-03-23T11:32:00.3638916Z             num_leaves=5,
2021-03-23T11:32:00.3639628Z             machines=",".join([
2021-03-23T11:32:00.3640331Z                 "127.0.0.1:" + str(port)
2021-03-23T11:32:00.3641301Z                 for port in open_ports
2021-03-23T11:32:00.3645753Z             ]),
2021-03-23T11:32:00.3646302Z         )
2021-03-23T11:32:00.3646761Z     
2021-03-23T11:32:00.3647176Z         dask_model2.fit(dX, dy, group=dg)
2021-03-23T11:32:00.3647895Z         assert dask_model2.fitted_
2021-03-23T11:32:00.3648388Z         params = dask_model2.get_params()
2021-03-23T11:32:00.3649217Z         assert 'local_listen_port' not in params
2021-03-23T11:32:00.3650342Z         assert 'machines' in params
2021-03-23T11:32:00.3651161Z     
2021-03-23T11:32:00.3651958Z         # model 3 - local_listen_port given
2021-03-23T11:32:00.3652472Z         # training should fail because LightGBM will try to use the same
2021-03-23T11:32:00.3653014Z         # port for multiple worker processes on the same machine
2021-03-23T11:32:00.3653488Z         dask_model3 = dask_model_factory(
2021-03-23T11:32:00.3653902Z             n_estimators=5,
2021-03-23T11:32:00.3654276Z             num_leaves=5,
2021-03-23T11:32:00.3654694Z             local_listen_port=listen_port
2021-03-23T11:32:00.3655075Z         )
2021-03-23T11:32:00.3655481Z         error_msg = "has multiple Dask worker processes running on it"
2021-03-23T11:32:00.3656048Z         with pytest.raises(lgb.basic.LightGBMError, match=error_msg):
2021-03-23T11:32:00.3656573Z >           dask_model3.fit(dX, dy, group=dg)
2021-03-23T11:32:00.3657297Z E           Failed: DID NOT RAISE <class 'lightgbm.basic.LightGBMError'>
2021-03-23T11:32:00.3657622Z 
2021-03-23T11:32:00.3658017Z ../tests/python_package_test/test_dask.py:1060: Failed
2021-03-23T11:32:00.3658754Z ---------------------------- Captured stderr setup -----------------------------
2021-03-23T11:32:00.3659810Z distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
2021-03-23T11:32:00.3660872Z distributed.scheduler - INFO - Clear task state
2021-03-23T11:32:00.3661642Z distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:44705
2021-03-23T11:32:00.3662618Z distributed.scheduler - INFO -   dashboard at:            127.0.0.1:8787
2021-03-23T11:32:00.3663603Z distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:37381
2021-03-23T11:32:00.3664626Z distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:37381
2021-03-23T11:32:00.3665468Z distributed.worker - INFO -          dashboard at:            127.0.0.1:43161
2021-03-23T11:32:00.3666187Z distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:44705
2021-03-23T11:32:00.3666860Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3667631Z distributed.worker - INFO -               Threads:                          1
2021-03-23T11:32:00.3668411Z distributed.worker - INFO -                Memory:                    8.35 GB
2021-03-23T11:32:00.3669360Z distributed.worker - INFO -       Local Directory: /__w/1/s/python-package/_test_worker-59578367-9a9e-4146-8046-12a28195c1e4/dask-worker-space/worker-sxa57r7x
2021-03-23T11:32:00.3670148Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3670830Z distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:39083
2021-03-23T11:32:00.3671559Z distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:39083
2021-03-23T11:32:00.3672321Z distributed.worker - INFO -          dashboard at:            127.0.0.1:45311
2021-03-23T11:32:00.3673083Z distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:44705
2021-03-23T11:32:00.3673740Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3674413Z distributed.worker - INFO -               Threads:                          1
2021-03-23T11:32:00.3675194Z distributed.worker - INFO -                Memory:                    8.35 GB
2021-03-23T11:32:00.3676065Z distributed.worker - INFO -       Local Directory: /__w/1/s/python-package/_test_worker-dab41a7a-2886-4e47-a534-0b90ff7ff829/dask-worker-space/worker-dedr43e2
2021-03-23T11:32:00.3677019Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3677927Z distributed.scheduler - INFO - Register worker <Worker 'tcp://127.0.0.1:37381', name: tcp://127.0.0.1:37381, memory: 0, processing: 0>
2021-03-23T11:32:00.3679032Z distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:37381
2021-03-23T11:32:00.3679706Z distributed.core - INFO - Starting established connection
2021-03-23T11:32:00.3680422Z distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:44705
2021-03-23T11:32:00.3681131Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3682110Z distributed.scheduler - INFO - Register worker <Worker 'tcp://127.0.0.1:39083', name: tcp://127.0.0.1:39083, memory: 0, processing: 0>
2021-03-23T11:32:00.3682928Z distributed.core - INFO - Starting established connection
2021-03-23T11:32:00.3683640Z distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:39083
2021-03-23T11:32:00.3684286Z distributed.core - INFO - Starting established connection
2021-03-23T11:32:00.3684967Z distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:44705
2021-03-23T11:32:00.3685675Z distributed.worker - INFO - -------------------------------------------------
2021-03-23T11:32:00.3686316Z distributed.core - INFO - Starting established connection
2021-03-23T11:32:00.3687035Z distributed.scheduler - INFO - Receive client connection: Client-c0e8c2b7-8bca-11eb-a01f-4f98255d250d
2021-03-23T11:32:00.3687807Z distributed.core - INFO - Starting established connection
2021-03-23T11:32:00.3688463Z ----------------------------- Captured stdout call -----------------------------
2021-03-23T11:32:00.3689032Z Finding random open ports for workers
2021-03-23T11:32:00.3689592Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3690055Z [LightGBM] [Info] Trying to bind port 42351...
2021-03-23T11:32:00.3690540Z [LightGBM] [Info] Binding port 42351 succeeded
2021-03-23T11:32:00.3691003Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3691495Z [LightGBM] [Warning] Connecting to rank 1 failed, waiting for 200 milliseconds
2021-03-23T11:32:00.3692052Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3692442Z [LightGBM] [Info] Trying to bind port 58255...
2021-03-23T11:32:00.3692858Z [LightGBM] [Info] Listening...
2021-03-23T11:32:00.3693235Z [LightGBM] [Info] Binding port 58255 succeeded
2021-03-23T11:32:00.3693609Z [LightGBM] [Info] Listening...
2021-03-23T11:32:00.3694004Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3694393Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3694818Z [LightGBM] [Info] Connected to rank 1
2021-03-23T11:32:00.3695278Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3695663Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3696104Z [LightGBM] [Info] Connected to rank 0
2021-03-23T11:32:00.3696761Z [LightGBM] [Info] Local rank: 0, total number of machines: 2
2021-03-23T11:32:00.3697289Z [LightGBM] [Info] Local rank: 1, total number of machines: 2
2021-03-23T11:32:00.3698123Z [LightGBM] [Warning] num_threads is set=1, n_jobs=-1 will be ignored. Current value: num_threads=1
2021-03-23T11:32:00.3698951Z [LightGBM] [Warning] num_threads is set=1, n_jobs=-1 will be ignored. Current value: num_threads=1
2021-03-23T11:32:00.3699687Z Using passed-in 'machines' parameter
2021-03-23T11:32:00.3700080Z [LightGBM] [Info] Listening...
2021-03-23T11:32:00.3700448Z [LightGBM] [Info] Listening...
2021-03-23T11:32:00.3700901Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3701366Z [LightGBM] [Warning] Set TCP_NODELAY failed
2021-03-23T11:32:00.3701989Z Using passed-in 'local_listen_port' for all workers
2021-03-23T11:32:00.3702654Z ----------------------------- Captured stderr call -----------------------------
2021-03-23T11:32:00.3703425Z distributed.worker - INFO - Run out-of-band function '_find_random_open_port'
2021-03-23T11:32:00.3704109Z distributed.worker - INFO - Run out-of-band function '_find_random_open_port'
2021-03-23T11:32:00.3704795Z --------------------------- Captured stderr teardown ---------------------------
2021-03-23T11:32:00.3705621Z distributed.scheduler - INFO - Remove client Client-c0e8c2b7-8bca-11eb-a01f-4f98255d250d
2021-03-23T11:32:00.3706382Z distributed.scheduler - INFO - Remove client Client-c0e8c2b7-8bca-11eb-a01f-4f98255d250d
2021-03-23T11:32:00.3707218Z distributed.scheduler - INFO - Close client connection: Client-c0e8c2b7-8bca-11eb-a01f-4f98255d250d
2021-03-23T11:32:00.3707967Z distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:39083
2021-03-23T11:32:00.3708716Z distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:37381
2021-03-23T11:32:00.3709706Z distributed.scheduler - INFO - Remove worker <Worker 'tcp://127.0.0.1:39083', name: tcp://127.0.0.1:39083, memory: 0, processing: 0>
2021-03-23T11:32:00.3710643Z distributed.core - INFO - Removing comms to tcp://127.0.0.1:39083
2021-03-23T11:32:00.3711570Z distributed.scheduler - INFO - Remove worker <Worker 'tcp://127.0.0.1:37381', name: tcp://127.0.0.1:37381, memory: 0, processing: 0>
2021-03-23T11:32:00.3712604Z distributed.core - INFO - Removing comms to tcp://127.0.0.1:37381
2021-03-23T11:32:00.3713283Z distributed.scheduler - INFO - Lost all workers
2021-03-23T11:32:00.3713862Z distributed.scheduler - INFO - Scheduler closing...
2021-03-23T11:32:00.3714539Z distributed.scheduler - INFO - Scheduler closing all comms
```

**Comment by StrikerRUS** [comment]
> One more time:

Created #4009.

**Pull request merged**